velox/common/memory/MemoryAllocator.cpp

```diff
@@ -191,34 +221,61 @@ bool MemoryAllocator::allocateContiguous(
     ContiguousAllocation& allocation,
     ReservationCallback reservationCB,
     MachinePageCount maxPages) {
-  if (cache() == nullptr) {
-    return allocateContiguousWithoutRetry(
-        numPages, collateral, allocation, reservationCB, maxPages);
-  }
-  auto numCollateralPages =
+  const auto numCollateralPages =
       allocation.numPages() + (collateral ? collateral->numPages() : 0);
-  const bool success = cache()->makeSpace(
-      pagesToAcquire(numPages, numCollateralPages), [&](Allocation& acquired) {
-        freeNonContiguous(acquired);
-        return allocateContiguousWithoutRetry(
-            numPages, collateral, allocation, reservationCB, maxPages);
-      });
-  if (!success) {
-    // There can be a failure where allocation was never called because there
-    // never was a chance based on numAllocated() and capacity(). Make sure old
-    // data is still freed.
-    int64_t freedBytes{0};
+  const auto totalCollateralBytes =
+      AllocationTraits::pageBytes(numCollateralPages);
+  const int64_t newPages = numPages - numCollateralPages;
+  auto cleanupCollateralAndReleaseReservation = [&](uint64_t reservationBytes) {
     if ((collateral != nullptr) && !collateral->empty()) {
-      freedBytes += AllocationTraits::pageBytes(collateral->numPages());
       freeNonContiguous(*collateral);
     }
     if (!allocation.empty()) {
-      freedBytes += allocation.size();
       freeContiguous(allocation);
     }
-    if ((reservationCB) != nullptr && (freedBytes > 0)) {
-      reservationCB(freedBytes, false);
+    if ((reservationCB) != nullptr && (reservationBytes > 0)) {
+      reservationCB(reservationBytes, false);
     }
+  };
+
+  if (numPages == 0) {
+    cleanupCollateralAndReleaseReservation(totalCollateralBytes);
```

Comment:
```
s/cleanupCollateralAndReleaseReservation/cleanupOnFailureCb/
```

Reply:
I avoided mentioning 'failure' in the lambda's name since the cleanup is not only used for failure cases but also used for the success case where numPages==0. Let me know if you feel strongly in favor of 'cleanupOnFailureCb ' and I can make the change